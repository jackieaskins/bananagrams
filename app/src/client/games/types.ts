import { BoardLocation } from "../../types/board";
import { Game } from "../../types/game";

export type GameState = {
  gameInfo: Game;
  handleDump: (tile: {
    id: string;
    boardLocation: BoardLocation | null;
  }) => void;
  handleMoveTileFromHandToBoard: (
    tileId: string,
    boardLocation: BoardLocation,
  ) => void;
  handleMoveTileFromBoardToHand: (boardLocation: BoardLocation) => void;
  handleMoveTileOnBoard: (
    fromLocation: BoardLocation,
    toLocation: BoardLocation,
  ) => void;
  handlePeel: () => void;
  isInGame: boolean;
};

export type GameLocationState = Partial<GameState>;
