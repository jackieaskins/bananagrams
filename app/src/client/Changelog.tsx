import {
  Container,
  Heading,
  Table,
  TableContainer,
  Tbody,
  Td,
  Th,
  Thead,
  Tr,
} from "@chakra-ui/react";

type Change = {
  id: string;
  date: string;
  description: string;
};

const DATE_FORMATTER = new Intl.DateTimeFormat("en-us", {
  dateStyle: "medium",
  timeZone: "UTC",
});

const MAJOR_CHANGES: Change[] = [
  {
    id: "redesign",
    date: "2024-01-18",
    description:
      "Gameplay has been completely redesigned! Give it a try, report bugs, and provide feedback.",
  },
  {
    id: "swap-tiles",
    date: "2023-12-02",
    description:
      "Added the ability to swap tiles on the board by placing one tile on top of another.",
  },
  {
    id: "leave-game-tiles",
    date: "2023-11-30",
    description:
      "Tiles will no longer return to the bunch when a player leaves the game.",
  },
  {
    id: "game-start",
    date: "2023-11-25",
    description:
      "Added start game button. The game will no longer automatically start when all players are ready.",
  },
  {
    id: "spectator-view",
    date: "2023-11-24",
    description:
      "Added spectator mode. Players can now join in-progress games as spectators!",
  },
  {
    id: "changelog",
    date: "2023-11-17",
    description: "Added changelog.",
  },
  {
    id: "dark-mode",
    date: "2023-11-11",
    description: "Added dark mode.",
  },
  {
    id: "and-we-back",
    date: "2023-11-10",
    description: "We're back!",
  },
].map(({ date, ...rest }) => ({
  date: DATE_FORMATTER.format(new Date(date)),
  ...rest,
}));

export default function Changelog(): JSX.Element {
  return (
    <Container>
      <Heading as="h1" textAlign="center" marginY="35px">
        Changelog
      </Heading>

      <TableContainer>
        <Table variant="striped">
          <Thead>
            <Tr>
              <Th textAlign="right">Date</Th>
              <Th>Description</Th>
            </Tr>
          </Thead>

          <Tbody>
            {MAJOR_CHANGES.map(({ id, date, description }) => (
              <Tr key={id}>
                <Td
                  textAlign="right"
                  overflowWrap="break-word"
                  verticalAlign="top"
                >
                  {date}
                </Td>
                <Td whiteSpace="normal">{description}</Td>
              </Tr>
            ))}
          </Tbody>
        </Table>
      </TableContainer>
    </Container>
  );
}
