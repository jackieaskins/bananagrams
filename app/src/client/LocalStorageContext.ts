import { createContext, useContext } from "react";
import { SetState } from "./state/types";

type LocalStorageState<T> = [T, SetState<T>];
type LocalStorageContextState = {
  gameSettings: {
    enableTileSwap: LocalStorageState<boolean>;
  };
  savedUsername: LocalStorageState<string>;
  showTutorialPrompt: LocalStorageState<boolean>;
};

export const LocalStorageContext = createContext<LocalStorageContextState>(
  {} as LocalStorageContextState,
);

function useLocalStorageContext(): LocalStorageContextState {
  return useContext(LocalStorageContext);
}

export function useEnableTileSwap(): LocalStorageState<boolean> {
  return useLocalStorageContext().gameSettings.enableTileSwap;
}

export function useSavedUsername(): LocalStorageState<string> {
  return useLocalStorageContext().savedUsername;
}

export function useShowTutorialPrompt(): LocalStorageState<boolean> {
  return useLocalStorageContext().showTutorialPrompt;
}
