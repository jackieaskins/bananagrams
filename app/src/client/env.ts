export const { DEV, PROD } = import.meta.env;
