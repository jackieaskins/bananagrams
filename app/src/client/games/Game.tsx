import { Box, Flex, useBreakpointValue } from "@chakra-ui/react";
import Konva from "konva";
import {
  useCallback,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from "react";
import GameFooter from "./GameFooter";
import GameSidebar from "./GameSidebar";
import Canvas from "@/client/canvas/Canvas";
import { CanvasContext } from "@/client/canvas/CanvasContext";
import KeysProvider from "@/client/keys/KeysProvider";
import { useNavMenu } from "@/client/menus/NavMenuContext";
import SelectedTilesProvider from "@/client/tiles/SelectedTilesProvider";

const TILE_SIZE = 32;

export default function Game(): JSX.Element {
  const gameSidebarRef = useRef<HTMLDivElement>(null);
  const gameBarRef = useRef<HTMLDivElement>(null);

  const [sidebarExpanded, setSidebarExpanded] = useState(true);
  const [hasManuallyToggledSidebar, setHasManuallyToggledSidebar] =
    useState(false);
  const shouldDefaultSidebarOpen = useBreakpointValue(
    { base: true, sm: false, md: true },
    { fallback: "md" },
  );
  const expandSidebar = useCallback((expanded: boolean) => {
    setSidebarExpanded(expanded);
    setHasManuallyToggledSidebar(true);
  }, []);
  useEffect(() => {
    if (!hasManuallyToggledSidebar) {
      setSidebarExpanded(shouldDefaultSidebarOpen!);
    }
  }, [hasManuallyToggledSidebar, shouldDefaultSidebarOpen]);

  const [cursorPosition, setCursorPosition] = useState({ x: 0, y: 0 });
  const [offset, setOffset] = useState({ x: 0, y: 0 });
  const stageRef = useRef<Konva.Stage>(null);
  const [size, setSize] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
  });

  const [, setRenderNavMenu] = useNavMenu();
  useEffect(() => {
    setRenderNavMenu(false);

    return () => {
      setRenderNavMenu(true);
    };
  }, [setRenderNavMenu]);

  useLayoutEffect(() => {
    const currGameBarRef = gameBarRef.current;
    const currGameSidebarRef = gameSidebarRef.current;

    function updateSize() {
      setSize({
        width: window.innerWidth - (currGameSidebarRef?.clientWidth ?? 0),
        height: window.innerHeight - (currGameBarRef?.clientHeight ?? 0),
      });
    }
    const observer = new ResizeObserver(updateSize);

    window.addEventListener("resize", updateSize);
    if (currGameBarRef) observer.observe(currGameBarRef);
    if (currGameSidebarRef) observer.observe(currGameSidebarRef);

    updateSize();

    return () => {
      window.removeEventListener("resize", updateSize);
      observer.disconnect();
    };
  }, []);

  return (
    <Flex height="100vh" width="100vw" direction="row">
      <Box>
        <CanvasContext.Provider
          value={{
            cursorPosition,
            offset,
            playable: true,
            size,
            stageRef,
            tileSize: TILE_SIZE,
          }}
        >
          <SelectedTilesProvider>
            <KeysProvider>
              <Canvas
                setOffset={setOffset}
                setCursorPosition={setCursorPosition}
              />
            </KeysProvider>
          </SelectedTilesProvider>
        </CanvasContext.Provider>

        <Box ref={gameBarRef}>
          <GameFooter sidebarExpanded={sidebarExpanded} />
        </Box>
      </Box>

      <Box ref={gameSidebarRef}>
        <GameSidebar expanded={sidebarExpanded} setExpanded={expandSidebar} />
      </Box>
    </Flex>
  );
}
