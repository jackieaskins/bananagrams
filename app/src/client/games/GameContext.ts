import { createContext, useContext } from "react";
import { GameState } from "./types";
import { Game } from "@/types/game";

export function getEmptyGameInfo(gameId: string): Game {
  return {
    gameId,
    gameName: "",
    isInProgress: false,
    bunch: [],
    players: [],
    previousSnapshot: null,
  };
}

export const GameContext = createContext<GameState>({} as GameState);

export function useGame(): GameState {
  return useContext(GameContext);
}
