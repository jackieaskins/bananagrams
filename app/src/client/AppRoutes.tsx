import { Route, Routes } from "react-router-dom";
import Changelog from "./Changelog";
import NotFound from "./NotFound";
import CreateGame from "./games/CreateGame";
import Game from "./games/Game";
import GameManager from "./games/GameManager";
import JoinGame from "./games/JoinGame";
import SocketGameProvider from "./games/SocketGameProvider";
import GameRedesign from "./redesign/GameRedesign";
import useSocketManager from "./socket/useSocketManager";
import Tutorial from "./tutorial/Tutorial";

function getGameRoutes(topLevelPath: string, game: JSX.Element) {
  return (
    <>
      <Route path={topLevelPath} element={<CreateGame />} />
      <Route
        path={`${topLevelPath}game/:gameId`}
        element={
          <SocketGameProvider>
            <GameManager game={game} />
          </SocketGameProvider>
        }
      />
      <Route path={`${topLevelPath}game/:gameId/join`} element={<JoinGame />} />
    </>
  );
}

export default function AppRoutes(): JSX.Element {
  useSocketManager();

  return (
    <Routes>
      {getGameRoutes("/", <Game />)}
      {getGameRoutes("/redesign/", <GameRedesign />)}
      <Route path="/tutorial" element={<Tutorial />} />
      <Route path="/changelog" element={<Changelog />} />
      <Route path="*" element={<NotFound />} />
    </Routes>
  );
}
