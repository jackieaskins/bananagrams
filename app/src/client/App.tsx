import {
  ChakraProvider,
  Spinner,
  Text,
  ThemeConfig,
  ToastProviderProps,
  extendTheme,
} from "@chakra-ui/react";
import { useEffect, useState } from "react";
import { BrowserRouter } from "react-router-dom";
import AppRoutes from "./AppRoutes";
import LocalStorageProvider from "./LocalStorageProvider";
import BraveBrowserDialog from "./dialogs/BraveBrowserDialog";
import ServerDisconnectDialog from "./dialogs/ServerDisconnectDialog";
import CenteredLayout from "./layouts/CenteredLayout";
import NavMenu from "./menus/NavMenu";
import { NavMenuContext } from "./menus/NavMenuContext";
import { socket } from "./socket";

const config: ThemeConfig = {
  initialColorMode: "system",
  useSystemColorMode: true,
};
const theme = extendTheme({ config });

const toastOptions: ToastProviderProps = {
  defaultOptions: {
    position: "top-left",
  },
};

export default function App(): JSX.Element {
  const [renderNavMenu, setRenderNavMenu] = useState(true);
  const [hasConnected, setHasConnected] = useState(socket.connected);

  useEffect(() => {
    function handleConnect() {
      setHasConnected(true);
    }

    socket.on("connect", handleConnect);
    socket.connect();

    return () => {
      socket.off("connect", handleConnect);
    };
  }, []);

  return (
    <LocalStorageProvider>
      <ChakraProvider theme={theme} toastOptions={toastOptions}>
        {hasConnected ? (
          <NavMenuContext.Provider value={[renderNavMenu, setRenderNavMenu]}>
            <BrowserRouter>
              <ServerDisconnectDialog />
              <AppRoutes />
              {renderNavMenu && <NavMenu />}
              <BraveBrowserDialog />
            </BrowserRouter>
          </NavMenuContext.Provider>
        ) : (
          <CenteredLayout>
            <Text textAlign="center">
              <Spinner />
            </Text>
          </CenteredLayout>
        )}
      </ChakraProvider>
    </LocalStorageProvider>
  );
}
