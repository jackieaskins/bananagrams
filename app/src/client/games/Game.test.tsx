import { shallow } from "enzyme";
import Game from "./Game";
import { useGame } from "./GameContext";
import { playerFixture } from "@/client/fixtures/player";
import { CURRENT_PLAYER_ID } from "@/client/testUtils";
import { PlayerStatus } from "@/types/player";

const mockUseGame = useGame as jest.Mock;
jest.mock("./GameContext", () => ({
  useGame: jest.fn(),
}));

jest.mock("../boards/validate", () => ({
  isValidConnectedBoard: () => false,
}));

describe("<Game />", () => {
  const renderComponent = () => shallow(<Game />);

  beforeEach(() => {
    mockUseGame.mockReturnValue({
      gameInfo: {
        bunch: [],
        players: [playerFixture({ userId: CURRENT_PLAYER_ID })],
      },
      handlePeel: jest.fn().mockName("handlePeel"),
    });
  });

  test("renders properly", () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  it("renders properly with more than one active player", () => {
    mockUseGame.mockReturnValue({
      gameInfo: {
        bunch: [],
        players: [
          playerFixture({
            userId: CURRENT_PLAYER_ID,
            status: PlayerStatus.READY,
          }),
          playerFixture({
            userId: "inactive",
            status: PlayerStatus.SPECTATING,
          }),
          playerFixture({ userId: "other", status: PlayerStatus.READY }),
        ],
      },
      handlePeel: jest.fn().mockName("handlePeel"),
    });

    expect(renderComponent()).toMatchSnapshot();
  });
});
