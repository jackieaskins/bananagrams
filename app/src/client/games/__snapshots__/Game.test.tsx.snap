// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> renders properly 1`] = `
<DndProvider
  backend={[Function]}
>
  <Center
    margin="50px"
  >
    <Stack
      alignItems={
        {
          "base": "center",
          "lg": "start",
        }
      }
      direction={
        {
          "base": "column",
          "lg": "row",
        }
      }
      justifyContent="center"
      spacing={8}
    >
      <Stack>
        <Text
          textAlign="center"
        >
          Your board and hand:
        </Text>
        <HStack>
          <Board
            board={{}}
          />
          <Hand
            hand={[]}
          />
        </HStack>
      </Stack>
      <Stack
        direction={
          {
            "base": "row",
            "lg": "column",
          }
        }
        flex="1"
      >
        <Stack>
          <Text
            textAlign="center"
          >
            Tiles remaining in bunch: 
            0
          </Text>
          <PeelButton />
          <Dump />
          <SpectateButton />
          <GameSettings />
        </Stack>
      </Stack>
    </Stack>
  </Center>
</DndProvider>
`;

exports[`<Game /> renders properly with more than one active player 1`] = `
<DndProvider
  backend={[Function]}
>
  <Center
    margin="50px"
  >
    <Stack
      alignItems={
        {
          "base": "center",
          "lg": "start",
        }
      }
      direction={
        {
          "base": "column",
          "lg": "row",
        }
      }
      justifyContent="center"
      spacing={8}
    >
      <Stack>
        <Text
          textAlign="center"
        >
          Your board and hand:
        </Text>
        <HStack>
          <Board
            board={{}}
          />
          <Hand
            hand={[]}
          />
        </HStack>
      </Stack>
      <Stack
        direction={
          {
            "base": "row",
            "lg": "column",
          }
        }
        flex="1"
      >
        <Stack>
          <Text
            textAlign="center"
          >
            Tiles remaining in bunch: 
            0
          </Text>
          <PeelButton />
          <Dump />
          <SpectateButton />
          <GameSettings />
        </Stack>
        <Stack
          alignItems="center"
        >
          <Text
            textAlign="center"
          >
            Opponent board(s):
          </Text>
          <OpponentBoardPreview
            players={
              [
                {
                  "board": {},
                  "gamesWon": 0,
                  "hand": [],
                  "isAdmin": false,
                  "isTopBanana": false,
                  "status": "READY",
                  "userId": "currentPlayerId",
                  "username": "username",
                },
                {
                  "board": {},
                  "gamesWon": 0,
                  "hand": [],
                  "isAdmin": false,
                  "isTopBanana": false,
                  "status": "READY",
                  "userId": "other",
                  "username": "username",
                },
              ]
            }
          />
        </Stack>
      </Stack>
    </Stack>
  </Center>
</DndProvider>
`;
